servant-server/test/Servant/ServerSpec.hs

```diff
@@ -574,3 +638,127 @@ miscReqCombinatorsSpec = with (return $ serve miscApi miscServ) $
       go "/host" "\"0.0.0.0:0\""
 
   where go path res = Test.Hspec.Wai.get path `shouldRespondWith` res
+
+
+-- | we include two endpoints /foo and /bar and we put the BasicAuth
+-- portion in two different places
+type AuthUser = ByteString
+type BasicAuthFooRealm = AuthProtect "foo" (BasicAuth "foo-realm") AuthUser 'Strict () 'Strict ()
+type BasicAuthBarRealm = AuthProtect "bar" (BasicAuth "bar-realm") AuthUser 'Strict () 'Strict ()
+type BasicAuthRequiredAPI = BasicAuthFooRealm :> "foo" :> Get '[JSON] Person
+                  :<|> "bar" :> BasicAuthBarRealm :> Get '[JSON] Animal
+
+basicAuthFooCheck :: BasicAuth "foo-realm" -> IO (Maybe AuthUser)
+basicAuthFooCheck (BasicAuth user pass) = if user == "servant" && pass == "server"
+                                          then return (Just "servant")
+                                          else return Nothing
+
+basicAuthBarCheck :: BasicAuth "bar-realm" -> IO (Maybe AuthUser)
+basicAuthBarCheck (BasicAuth usr pass) = if usr == "bar" && pass == "bar"
+                                         then return (Just "bar")
+                                         else return Nothing
+
+basicBasicAuthRequiredApi :: Proxy BasicAuthRequiredAPI
+basicBasicAuthRequiredApi = Proxy
+
+basicAuthRequiredServer :: Server BasicAuthRequiredAPI
+basicAuthRequiredServer = (const . return $ alice)
+                :<|> (const . return $ jerry)
+
+-- base64-encoded "servant:server"
+base64ServantColonServer :: ByteString
+base64ServantColonServer = "c2VydmFudDpzZXJ2ZXI="
+
+-- base64-encoded "bar:bar"
+base64BarColonPassword :: ByteString
+base64BarColonPassword = "YmFyOmJhcg=="
+
+-- base64-encoded "user:password"
+base64UserColonPassword :: ByteString
+base64UserColonPassword = "dXNlcjpwYXNzd29yZA=="
+
+basicAuthGet :: ByteString -> ByteString -> WaiSession SResponse
+basicAuthGet path base64EncodedAuth = Test.Hspec.Wai.request methodGet path [("Authorization", "Basic " <> base64EncodedAuth)] ""
+
+basicAuthRequiredSpec :: Spec
+basicAuthRequiredSpec = do
+    describe "Servant.API.Authentication" $ do
+        let fooAuthProtect = basicAuthStrict basicAuthFooCheck
+            barAuthProtect = basicAuthStrict basicAuthBarCheck
+            config :: Config [ConfigEntry "foo" (AuthProtected IO ServantErr 'Strict () 'Strict () (BasicAuth "foo-realm") AuthUser), ConfigEntry "bar" (AuthProtected IO ServantErr 'Strict () 'Strict () (BasicAuth "bar-realm") AuthUser)]
+            config = fooAuthProtect .: barAuthProtect .: EmptyConfig
+        with (return $ serve basicBasicAuthRequiredApi config basicAuthRequiredServer) $ do
+            it "allows access with the correct username and password" $ do
```

Comment:
@jkarni above is some example usage of `Config` at the `server` level. This example is using the more complicated `AuthProtect`